opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+		if (k1.size() == 1 && 2 == k2.size() && k1[0] > k2[0] && k1[0] < k2[1])
+			continue;
+
+		//Singleton Interval of Key1 isn't contained in Interval of Key2
+		if (k1.size() == 1 && 2 == k2.size() && (k1[0] < k2[0] || k1[0] > k2[1]))
+			return 0;
+
+		// The part of an interval contained in a sigelton set is infinitelly small
+		if (k1.size() == 2 && 1 == k2.size())
+			return 0;
+
+		//Two Intervals
+		//Calculate the overlapp
+		Interval res;
+		//Find the start of the Overlapp
+		if (k1[0] > k2[0])
+			res.push_back(k1[0]);
+		else
+			res.push_back(k2[0]);
+
+		//Find the end of the Overlapp
+		if (k1[1] < k2[1])
+			res.push_back(k1[1]);
+		else
+			res.push_back(k2[1]);
+
+		//Check that the Overlapp starts before it ends
+		if (res[0] <= res[1])
+			//Figure out how much smaler the Overlapp is then K1
+			sum *= (res[1] - res[0]) / (k1[1] - k1[0]);
+		else
+			return 0;
+	}
+	return sum;
+}
+
+//Get the Count of a Key that might not be in the DV explicitly
+//by a weighted sum of all the Keys that are in the DV
+//weighted by the overlapp of the given Key with the Keys of the DV
+double DistributionalValue::get_contained_count(DVKey h) const
+{
+	double res = 0;
+	for (auto v : _value)
+	{
+		double weigth = DistributionalValue::key_contained(h,v.first);
+		res += v.second * weigth;
+	}
+	return res;
+}
+
+double DistributionalValue::get_mode(DVKey val) const
+{
+	return get_mode_for(get_count(val));
+}
+double DistributionalValue::get_mean(DVKey val) const
+{
+	return get_mean_for(get_count(val));
+}
+double DistributionalValue::get_contained_mean(DVKey val) const
+{
+	return get_mean_for(get_contained_count(val));
+}
+double DistributionalValue::get_var(DVKey val) const
+{
+	return get_var_for(get_count(val));
+}
+
+std::string DistributionalValue::to_string(const std::string& indent) const
+{
+	std::stringstream ss;
+	if (_value.size() == 0)
+		ss << "Empty DistributionalValue" << std::endl;
+	for (auto elem : _value)
+	{
+		ss << indent << "{";
+		for (auto interval : elem.first)
+		{
+			if (interval.size() == 1)
+				ss << interval[0] << ";";
+			else
+				ss << "["
+				   << interval[0]
+				   << ","
+				   << interval[1]
+				   << ");";
+		}
+		ss.seekp(-1,std::ios_base::end);
+		ss << "}"
+		   << " Count: "
+		   << std::setprecision(18)
+		   << elem.second
+		   << " Mean: "
+		   << get_mean_for(elem.second)
+		   << std::endl;
+	}
+	return ss.str();
+}
+
+bool DistributionalValue::operator==(const Value& other) const
+{
+	if (DISTRIBUTIONAL_VALUE != other.get_type()) return false;
+	const DistributionalValue* dov = (const DistributionalValue*) &other;
+
+	if (_value.size() != dov->_value.size()) return false;
+
+	for (auto elem : _value) {
+		double v1 = elem.second;
+		double v2 = dov->_value.get(elem.first);
+		if (not is_approx_eq_ulp(v1,v2,24))
+			return false;
+	}
+	return true;
+}
+
+std::string oc_to_string(const DistributionalValuePtr dvp)
+{
+	std::string part = dvp->to_string("");
+	return part;
+}
```

Comment:
1. `oc_to_string` now always takes an optional indent argument. I had done 2 versions, one without indent and one with indent, to work around some gcc/gdb issue, but it's been fixed in gcc 5.
2. Why not just `return dvp->to_string(indent);` instead of creating the intermediary variable `part`?